### PR TITLE
add cwd aware hinter

### PR DESCRIPTION
**Pull request by p00f**
towards https://github.com/nushell/nushell/issues/8883

### CODE REVIEW

**Comment by p00f** [comment]
I tried this out and it works - and I love the experience. This was the only thing I missed from fish.

Help with writing tests would be appreciated

**Comment by fdncred** [comment]
Nice! It would be good to have an example that demonstrates its use in the examples folder. Can you add one of those please?

**Comment by p00f** [comment]
> Nice! It would be good to have an example that demonstrates its use in the examples folder. Can you add one of those please?

I tried running the existing one (`cargo run --example hinter`) and don't actually see hints - do you want me to just copy-paste it into a new example and just change the constructor like so:

```rust

// Create a reedline object with in-line hint support.
// cargo run --example cwd_aware_hinter
//
// Fish-style cwd history based hinting
// assuming history ["abc", "ade"]
// pressing "a" hints to abc.
// Up/Down or Ctrl p/n, to select next/previous match

use nu_ansi_term::{Color, Style};
use reedline::{CwdAwareHinter, DefaultPrompt, Reedline, Signal};
use std::io;

fn main() -> io::Result<()> {
    let mut line_editor = Reedline::create().with_hinter(Box::new(
        CwdAwareHinter::default().with_style(Style::new().italic().fg(Color::LightGray)),
    ));
    let prompt = DefaultPrompt::default();

    loop {
        let sig = line_editor.read_line(&prompt)?;
        match sig {
            Signal::Success(buffer) => {
                println!("We processed: {buffer}");
            }
            Signal::CtrlD | Signal::CtrlC => {
                println!("\nAborted!");
                break Ok(());
            }
        }
    }
}
```

**Comment by fdncred** [comment]
I think so, but you'll need to add some items to the history in order to get hints to work.

**Comment by p00f** [comment]
right - how do I change the cwd inside the repl though? One idea is to alternate between original cwd and `std::env::home_dir()` on every iteration.

**Comment by p00f** [comment]
How does the example look? I tested the example and it works as intended.

**Comment by p00f** [comment]
I think I need help writing tests as I'm not familiar with the code too much and there are no tests for the existing`DefaultHinter`

**Comment by p00f** [comment]
On the nushell side I just use `DefaultHinter` when the file format is `FileBackedHistory` (I'm assuming `engine_state` is updated at runtime, so the behaviour is that it is just not used)

**Comment by p00f** [comment]
https://paste.sr.ht/~p00f/5752fc99ba31990710f08c2a85de897a5b15bf54#cwd.diff-L48

@sholderbach this is my nushell patch

**Comment by p00f** [comment]
Note that this causes no undesirable behaviour - if you change the history format you don't get hints from the other history file anyway

**Comment by p00f** [comment]
> (I'm assuming engine_state is updated at runtime, so the behaviour is that it is just not used)

this is correct - the `Hinter` is created [inside](https://github.com/nushell/nushell/blob/5d8763ed1d14048ac9fc09eb21be36e6e622f0d7/crates/nu-cli/src/repl.rs#L302) the [REPL loop](https://github.com/nushell/nushell/blob/5d8763ed1d14048ac9fc09eb21be36e6e622f0d7/crates/nu-cli/src/repl.rs#L193)

**Comment by sholderbach** [comment]
My caution is driven by the fact that we have a [substantial number of non-nushell users](https://github.com/nushell/reedline/network/dependents) of reedline. And since we introduced the new history search API (#401) we had a few panics based on the capabilities or assumptions what info is accessible (e.g. #480).

I would like to avoid a situation where you would encounter runtime panics when the `Hinter`/`History` combination gets used.

Things should be either abundantly clear from documentation, failure should happen either at the earliest point possible or handled gracefully.

> [paste.sr.ht/~p00f/5752fc99ba31990710f08c2a85de897a5b15bf54#cwd.diff-L48](https://paste.sr.ht/~p00f/5752fc99ba31990710f08c2a85de897a5b15bf54#cwd.diff-L48)

This would bring every user of `file_format: "sqlite"` over to the behavior that prioritizes hints for the current directory.
Not sure if this is universally liked. So there might be someone proposing a decoupling between the two down the line.

**Comment by p00f** [comment]
I see - there's no way to find out the type of `History`, the trait has no such methods. Do I create a `HistoryType` enum and the implement such a method? If I do end up doing this, why not just do away with dynamic dispatch and use enums instead?

imo I should just document that it should always be used with `SqliteBackedHistory`.

>  Not sure if this is universally liked. So there might be someone proposing a decoupling between the two down the line.

If they do, I'll add a config option. At least for now, I think everyone will love this - fish has this by default without a config option, nobody complains.

**Comment by fdncred** [comment]
> I think everyone will love this ...

Personally, I pass on this behavior. I'll play with it and test it out, but I don't think I'll be using it. Let's make it configurable. 

I think it's great to have it, and there's no doubt that some people will love it. However, I've worked on nushell long enough to know that you will never please everyone. 

On a separate note, maybe I'll change my nick from "fdncred" to "everyone" then we can list the features that 'everyone' loves. 🤣

**Comment by p00f** [comment]
> I'll play with it and test it out.

of course, I haven't created a nushell PR yet.

> On a separate note, maybe I'll change my nick from "fdncred" to "everyone" then we can list the features that 'everyone' loves. 🤣

Please do, I hate when people ping `@everyone` on discord - this way it won't actually ping everyone 😛

**Comment by fdncred** [comment]
> Please do, I hate when people ping @everyone on discord - this way it won't actually ping everyone.

LOL - me too. @here and @everyone kind of drive me crazy on discord.

File: src/hinter/cwd_aware.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
+#[cfg(any(feature = "sqlite", feature = "sqlite-dynlib"))]
+use crate::{history::SearchQuery, Hinter, History};
+#[cfg(any(feature = "sqlite", feature = "sqlite-dynlib"))]
+use nu_ansi_term::{Color, Style};
+
+/// A hinter that uses the completions or the history to show a hint to the user
+/// NOTE: Only use this with `SqliteBackedHistory`!
```
Comment (p00f) [inline src/hinter/cwd_aware.rs:7]: @sholderbach How does this look?

**Comment by sholderbach** [comment]
Briefly, I don't think just documenting it is a good approach for the public API.

With a `FileBackedHistory` things can be compiled, the builder executes and then things panic on the first keypress.
<details>
<summary>Panicking patch</summary>

```
cargo run --features sqlite --example cwd_aware_hinter
```

```patch
diff --git a/examples/cwd_aware_hinter.rs b/examples/cwd_aware_hinter.rs
index 17aa7dd..35f9c2f 100644
--- a/examples/cwd_aware_hinter.rs
+++ b/examples/cwd_aware_hinter.rs
@@ -49,16 +49,14 @@ fn main() -> io::Result<()> {
         #[allow(deprecated)]
         let home_dir = std::env::home_dir().unwrap();
 
-        let history = create_filled_example_history(
-            &home_dir.to_string_lossy().to_string(),
-            &orig_dir.to_string_lossy().to_string(),
-        );
+        let history = reedline::FileBackedHistory::new(100); 
+
 
         let mut line_editor = Reedline::create()
             .with_hinter(Box::new(
                 CwdAwareHinter::default().with_style(Style::new().italic().fg(Color::Yellow)),
             ))
-            .with_history(history);
+            .with_history(Box::new(history));
 
         let prompt = DefaultPrompt::default();
``` 
</details>

Either we replace the current `.expect()`s in a way we can implement fall backs through recoverable error handling or we block at construction in `.with_hinter`/`.with_history` and provide a recoverable error there.

### fallback/local error handling
As `History::search` already returns a result turning the expect here into a fall-back to a simpler query would be an option.
https://github.com/p00f/reedline/blob/00c59e4ca2762d5a12579b68de5dd505318fbfc4/src/hinter/cwd_aware.rs#L31

### check at construction time

You are right that trait objects are limited on its own, but for "RTTI we have at home" the cost of adding a method that advertises a particular capability doesn't sound too bad if it is only checked in the constructors.

**P.S.:** just to disagree with @fdncred I really like this feature as a default and would be ok with saving ourselves another config option (just strawmanning it in for API design purposes)

**Comment by p00f** [comment]
Sorry, it took me a while to realize how "just don't use it" sounds.

I went with the first option as the second one requires more thought. Thanks for the help!

**Comment by p00f** [comment]
re: config option, we can introduce this without an option on `main`, then add it if people want to disable it. There's some time before the next release.

**Comment by p00f** [comment]
@sholderbach i fixed the warning causing the CI failure, can you re-run it?

**Comment by sholderbach** [comment]
Thanks for implementing the fallback! 

> re: config option, we can introduce this without an option on `main`, then add it if people want to disable it. There's some time before the next release.

Agree, your nushell patch looked already pretty good.

**Review by sholderbach — APPROVED** [review]

**Comment by p00f** [comment]
sorry, I think this one should pass

**Comment by p00f** [comment]
should I squash?

**Comment by sholderbach** [comment]
No need for manual squashing, we happily use the Github button for that.

Thanks for improving the quality of the hints. Looking forward to using them in Nushell

**Comment by p00f** [comment]
How do I do the nushell PR? Uncommenting https://github.com/nushell/nushell/blob/d0dc6986dd14e7ea5443f3dc0eb5b101c5de5efc/Cargo.toml#L167 ?

**Comment by sholderbach** [comment]
Yup uncommenting that line and running `cargo update -p reedline` to guarantee you get the most recent version on that branch

**Comment by p00f** [comment]
Yeah - I was wondering if it was acceptable to commit into main, then I read the diff - it was uncommented in `main` when I started working  😄

**Pull request merged**